### ISSUE TITLE
Builds are slow when using multiple remote git repos, particular with "touch to sign"

### ISSUE THREAD

**Issue by pmembrey** [earthly/earthly#1023]
Hi guys,

Currently I have a project that has several github repositories for the project's dependencies, and every time I want to do a quick build to test a feature, Earthly checks each of these repositories to look for changes. This is relatively slow anyway (compared to not using remote git repos), but I also use touch to sign with a yubikey.

That means every time I do a build, I have to wait for the yubikey to flash, touch it, and then repeat for as many repos as I have. This makes a lightening fast sub second build take at least ten seconds with lots of manual input. This is made especially annoying because you have to stare at the yubikey as you don't know quite when it will start flashing.

Would it be possible to add a flag like `--git-no-update` or similar so that I can get super fast builds back?

Thanks! :)

**Comment by vladaionescu** [comment]
Hi @pmembrey - wondering what your exact setup is. Let me confirm that I understand correctly:

* You have an SSH key pair based on either ed25519-sk or ecdsa-sk and you've configured your local git access to use that?
* In the Earthfile you have remote references such as `COPY github.com/foo/bar+something/artifact`, but not `GIT CLONE`s?

If yes, are you using an SSH agent by any chance? Do you know if Earthly recognizes it correctly?

I'm not very familiar with "touch to sign" - tried googling what it means exactly, but came out empty. Maybe you can point me to some docs / guide so I can get a better idea?

A `--git-no-update` feature should be possible theoretically. It requires some buildkit changes though, so it might take a while.

**Comment by pmembrey** [comment]
Hi @vladaionescu , Earthly works perfectly with the Yubikey. It's detected just fine and is able to authenticate. Was impressed to see that work out of the box!

Touch to sign is where the device requires physically tapping it in order to get it to sign the request. So for example, when I do a `git pull`, the yubikey will start to flash to tell me something is trying to make a request. Once I tap it, it will complete the request and all is well. This is a security feature to prevent a remote attacker from being able to use a plugged in key to access remote systems, because they can't physically tap the key.

The problem is that it must (in my case) be tapped for every request. When I'm using multiple repos, this is problematic because every time I do a build, Earthly reaches out to see if any of the repos have changed. This means waiting on the Yubikey to start flashing and tapping it constantly to get through the check process.

Even without touch to sign, this is a slow process, making the build take 20 seconds instead of half a second. The beauty of Earthly is how trivially I can rebuild and test, but that feature is lost because of the time it takes to check the repos.

Touch to sign just makes this worse because of the constant interaction needed for every single request (and you have to tap it when it flashes so you have to stare at it and wait for the request - never knowing if you got them all).

Hope that explains it a bit better :)

**Comment by vladaionescu** [comment]
Got it - I'm wondering about the other details as well, so we can try to reproduce the issue exactly the way that you are experiencing it. Like what setup do you have for the SSH key pair, any SSH agent tweaks that may be needed, and what exactly you have in the Earthfile?

**Comment by pmembrey** [comment]
Sorry, I think I've given the wrong impression. Earthly is not slower than doing it just on my laptop. It's not the SSH key stuff that's causing the problem. The issue is that touch to sign means github requests become serialised, and I have to watch for the flashing light before I can touch the key. This process is always slow compared to say a simple SSH key, but because Earthly checks every repo every time (and I have many repos), it means I can't use Earrthly for quick builds or testing, because each time I run it I need to spend 20+ seconds going through the dance.

That said I can certainly give you the details :)

I'm using a series 5 Yubikey as my secure two factor:

https://www.yubico.com/products/yubikey-5-overview/

I'm using the GPG feature of the key, and `gpg-agent` to make a GPG key work with SSH. So basically I'm using a GPG key to authenticate to github. Again, this in itself isn't slow, but because I have touch to sign enabled, i.e. I need to physically touch the key, it becomes a much bigger pain.

However, even with a normal SSH key and SSH agent, the number of repos means my builds will take 10 - 15 seconds instead of completing in under a second. Earthly goes from so fast I don't need to think about it, to so slow that I have to change my "quick test" behaviour.

Hope this helps :)

**Comment by vladaionescu** [comment]
Ok so you're using the `gpg-agent` as the SSH agent. :+1:

I think that a `--git-no-update` feature should be possible. We have to make this change in buildkit, however, so it might take a while for us to get to this.

**Comment by vladaionescu** [comment]
Another idea - this might reduce the number of touches as well possibly: https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Multiplexing . But this needs to be integrated into buildkit too.

**Comment by pmembrey** [comment]
My key is set up to require a touch for every request. I don't think multiplexing would help with that problem, but it could certainly give a performance boost for repos using normal SSH authentication.

For me, I just want to avoid all of the round trips to github - even without the touch to sign, this is still rather slow. Again, no fault of Earthly, but it really slows down the build process down a lot.

**Comment by gilescope** [comment]
See the same thing with my yubikey. I have rust repos that reference other repos but I need to do a touch to pull. One option is to configure git so that that's not required, but when I do a normal git clone it's not required so something earthly does seems to require more privileges. Obviously configuring everyone's gits is a hassle. It would be nice if it worked out of the box.

So I would recommend that to solve this issue Earthly devs should be good open source devs and sign their commits with a yubikey or similar for added security that you're not being impersonated.

**Comment by nlewo** [comment]
@pmembrey I encounter the same issue and "fixed" it by using a Github Personnal Access Token  instead of the SSH auth. See https://docs.earthly.dev/docs/guides/auth#git-authentication

In order to have reproducible build file, I specify a Git commit ID when using `GIT CLONE`. But even when a Git commit is specified, earthly fetches the repository. I think it could knows the Git commit has already been fetched. Since a Git commit ID is a content address, it is not required to fetch it again.

However, i still have another issue related to yubikey: each time i run `earthly build` , my yubikey flashes and i don't know why yet. This even appends when the target doesn't exist...
